test/Dialect/FIRRTL/dedup.mlir

```diff
@@ -0,0 +1,411 @@
+// verification checking.
+// CHECK-LABEL: firrtl.circuit "Flip"
+firrtl.circuit "Flip" {
+  firrtl.module @Flip0(out %io: !firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>) {
+    %0 = firrtl.subfield %io(0) : (!firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>) -> !firrtl.uint<1>
+    %1 = firrtl.subfield %io(1) : (!firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>) -> !firrtl.uint<1>
+    firrtl.connect %1, %0 : !firrtl.uint<1>, !firrtl.uint<1>
+  }
+  firrtl.module @Flip1(out %io: !firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>) {
+    %0 = firrtl.subfield %io(0) : (!firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>) -> !firrtl.uint<1>
+    %1 = firrtl.subfield %io(1) : (!firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>) -> !firrtl.uint<1>
+    firrtl.connect %1, %0 : !firrtl.uint<1>, !firrtl.uint<1>
+  }
+  firrtl.module @Flip(out %io: !firrtl.bundle<foo: bundle<foo flip: uint<1>, fuzz: uint<1>>, bar: bundle<bar flip: uint<1>, buzz: uint<1>>>) {
+    %0 = firrtl.subfield %io(1) : (!firrtl.bundle<foo: bundle<foo flip: uint<1>, fuzz: uint<1>>, bar: bundle<bar flip: uint<1>, buzz: uint<1>>>) -> !firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>
+    %1 = firrtl.subfield %io(0) : (!firrtl.bundle<foo: bundle<foo flip: uint<1>, fuzz: uint<1>>, bar: bundle<bar flip: uint<1>, buzz: uint<1>>>) -> !firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>
+    %foo_io = firrtl.instance foo  @Flip0(out io: !firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>)
+    %bar_io = firrtl.instance bar  @Flip1(out io: !firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>)
+    firrtl.partialconnect %1, %foo_io : !firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>, !firrtl.bundle<foo flip: uint<1>, fuzz: uint<1>>
+    firrtl.connect %0, %bar_io : !firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>, !firrtl.bundle<bar flip: uint<1>, buzz: uint<1>>
+  }
+}
+
+
+// Don't attach empty annotations onto ops without annotations.
+// CHECK-LABEL: firrtl.circuit "NoEmptyAnnos"
+firrtl.circuit "NoEmptyAnnos" {
+  // CHECK-LABEL: @NoEmptyAnnos0()
+  firrtl.module @NoEmptyAnnos0() {
+    // CHECK: %w = firrtl.wire  : !firrtl.bundle<a: uint<1>>
+    // CHECK: %0 = firrtl.subfield %w(0) : (!firrtl.bundle<a: uint<1>>) -> !firrtl.uint<1>
+    %w = firrtl.wire : !firrtl.bundle<a: uint<1>>
+    %0 = firrtl.subfield %w(0) : (!firrtl.bundle<a: uint<1>>) -> !firrtl.uint<1>
+  }
+  firrtl.module @NoEmptyAnnos1() {
+    %w = firrtl.wire : !firrtl.bundle<a: uint<1>>
+    %0 = firrtl.subfield %w(0) : (!firrtl.bundle<a: uint<1>>) -> !firrtl.uint<1>
+  }
+  firrtl.module @NoEmptyAnnos() {
+    firrtl.instance empty0 @NoEmptyAnnos0()
+    firrtl.instance empty1 @NoEmptyAnnos1()
+  }
+}
+
+
+// Don't deduplicate modules with NoDedup.
+// CHECK-LABEL: firrtl.circuit "NoDedup"
+firrtl.circuit "NoDedup" {
+  firrtl.module @Simple() { }
+  // CHECK: firrtl.module @NoDedup 
+  firrtl.module @NoDedup() attributes {annotations = [{class = "firrtl.transforms.NoDedupAnnotation"}]} { }
+}
```

Comment:
Can this test be modified to place the `NoDedupAnnotation` or a submodule which would dedup with another submodule?

This test inadvertently hits a corner case where I don't think the main module is supposed to dedup with anything else. This is getting into the realm of pedantically-exact SFC behavior, but the SFC does do this. Consider:

```scala
circuit Foo:
  module Baz:
    input a: UInt<1>
    output b: UInt<1>

    b <= a

  module Bar:
    input a: UInt<1>
    output b: UInt<1>

    b <= a

  module Foo:
    input a: UInt<1>
    output b: UInt<1>

    b <= a
```

This will result in `Baz` and `Bar` deduping, but not with `Foo` using `firrtl -i Foo.fir -cll firrtl.transforms.DedupModules:trace`:

```scala
circuit Foo :
  module Baz :
    input a : UInt<1>
    output b : UInt<1>

    b <= a

  module Foo :
    input a : UInt<1>
    output b : UInt<1>

    b <= a
```

I would add a test that the main module can't dedup with anything, as well.

Reply:
So it turns out that this test is completely broken.

One thing we talked about in the past is that it should be *illegal* for any module to be unreachable from the top of the hierarchy.  This means that it should actually be impossible for the "main module" to dedup with anything, unless we had a recursive module instantiation (also illegal).  I'm not sure how hard we are on this stance, and we definitely never implemented any verification for this.

With that in mind, the dedup pass never actually sees modules that are unreachable.  This is because we visit modules leaf-first using `llvm::post_order(mainModule)`, which implicitly does not include unreachable modules.

So I have updated this test as you suggested above, a single top module instantiating two submodule.  I can add the second test suggested, and it would work because unreachable modules are never deduped with anything, or I can omit it because it's illegal.

Reply:
It's fine to just update the existing test. I just wanted to make sure that we weren't opening the door for deduplication with out-of-tree modules (which the test seemed to be relying on to work). 😉